shell/src/shell.rs

```diff
@@ -99,34 +85,121 @@ pub async fn run_forever(config: ShellConfig) -> color_eyre::Result<()> {
     Ok(())
 }
 
-pub async fn start_protocol<C, KBE>(
-    clients: Vec<C>,
+pub fn start_protocol_by_role<KBE>(
+    role: RoleType,
+    runtime: TangleRuntime,
     networks: Vec<GossipHandle>,
     account_id: sr25519::Public,
     logger: DebugLogger,
     pallet_tx: Arc<SubxtPalletSubmitter<TangleConfig, PairSigner<TangleConfig>>>,
     keystore: ECDSAKeyStore<KBE>,
-) -> color_eyre::Result<()>
+) -> color_eyre::Result<tokio::task::AbortHandle>
 where
-    C: ClientWithApi + 'static,
     KBE: KeystoreBackend,
 {
-    let node_input = NodeInput {
-        clients,
-        account_id,
-        logger,
-        pallet_tx,
-        keystore,
-        node_index: 0,
-        additional_params: (),
-        prometheus_config: PrometheusConfig::Disabled,
-        networks,
+    use RoleType::*;
+    use ThresholdSignatureRoleType::*;
+    let handle = match role {
+        Tss(DfnsCGGMP21Stark) | Tss(DfnsCGGMP21Secp256r1) | Tss(DfnsCGGMP21Secp256k1) => {
+            let node_input = NodeInput {
+                clients: vec![
+                    TangleRuntime::new(runtime.client()),
```

Comment:
Explain this more? Why do we need 4 runtimes?

Reply:
It is not 4 different connections, they all share the underlying websocket connection here. (Same subxt client). For why different instances and not only one, because we have four different protocols that pull the finality notification stream at the same time and that may block.
So we create 4 different streams that are basically reading from a single stream (subxt connection)